contracts/orderbook/Logic/OrderBookOrderPlace.sol

```diff
@@ -0,0 +1,258 @@
+                    IDeposits(vault).getTokenUsed(Order.orderID),
+                "OrderBook: invalid used token"
+            );
+            uint256 amountUsedOld = _allOrders[Order.orderID].loanTokenAmount +
+                _allOrders[Order.orderID].collateralTokenAmount;
+            if (amountUsedOld > amountUsed) {
+                IDeposits(vault).partialWithdraw(
+                    msg.sender,
+                    Order.orderID,
+                    amountUsedOld - amountUsed
+                );
+            } else {
+                IDeposits(vault).deposit(
+                    Order.orderID,
+                    amountUsed - amountUsedOld,
+                    msg.sender,
+                    usedToken
+                );
+            }
+        }
+        _allOrders[Order.orderID] = Order;
+        emit OrderAmended(
+            msg.sender,
+            Order.orderType,
+            Order.amountReceived,
+            Order.orderID,
+            Order.base,
+            Order.loanTokenAddress
+        );
+    }
+
+    function cancelOrder(bytes32 orderID) external pausable {
+        require(!_allOrders[orderID].isCancelled, "OrderBook: inactive order");
+        _allOrders[orderID].isCancelled = true;
+        require(_histOrders[msg.sender].remove(orderID), "OrderBook: not owner of order");
+        _allOrderIDs.remove(orderID);
+        if (_allOrders[orderID].orderType == IOrderBook.OrderType.LIMIT_OPEN) {
+            address usedToken = IDeposits(vault).getTokenUsed(
+                orderID
+            );
+            IDeposits(vault).withdrawToTrader(msg.sender, orderID);
+        }
+        emit OrderCancelled(msg.sender, orderID);
+    }
+
+    function recoverFundsFromFailedOrder(bytes32 orderID) external pausable {
+        IOrderBook.Order memory order = _allOrders[orderID];
+        require(msg.sender == order.trader, "OrderBook: Not trade owner");
+        require(order.isCancelled, "OrderBook: Order not executed");
+        require(!_allOrderIDs.contains(orderID), "OrderBook: Order still in records");
+        IDeposits(vault).withdrawToTrader(msg.sender, orderID);
+    }
+
+    function cancelOrderProtocol(bytes32 orderID) external pausable {
+        IOrderBook.Order memory order = _allOrders[orderID];
+        address trader = order.trader;
+        require(!order.isCancelled, "OrderBook: inactive order");
+        uint256 amountUsed = order.collateralTokenAmount +
+            order.loanTokenAmount;
+        uint256 swapRate;
+        if (order.orderType == IOrderBook.OrderType.LIMIT_OPEN) {
+            swapRate = queryRateReturn(
+                order.loanTokenAddress,
+                order.base,
+                amountUsed
+            );
+        } else {
+            swapRate = queryRateReturn(
+                order.base,
+                order.loanTokenAddress,
+                amountUsed
+            );
+        }
+        require(
+            (
+                order.amountReceived > swapRate
```

Comment:
how are  you comparing swapRate with amount ?

Reply:
this is just to see which is greater so I do not have to deal with negatives